### ISSUE TITLE
Explain

### ISSUE THREAD

**Issue by MagoKimbra** [zodier/MFRC522-UART-Arduino#1]
Please if possible explain me this library.. I have by module RC522 width UART...
Sorry for my english!!!

**Comment by zodier** [comment]
Hi MagoKimbra, you have a simple example in the "examples" folder. If you have any specific question, please, ask here and I will try to help you. :)

**Comment by jr-k** [comment]
I set the MISO(TX) pin on D0 and SDA(RX) on D1 but I got nothing... I have the RFID-RC522, do I need to connect 2 pins to activate UART mode ? 

Also I've tested my device work with SPI and it's working.

Can you help me pls ?

**Comment by zodier** [comment]
Hi jreziga, what model of Arduino are you using? Note that some models of Arduino share the Serial Bus (like Arduino nano) with the USB connector. While you are using the USB connection with your computer the Serial Bus with RC522 doesn't works fine. If you are using an Arduino Mega, please use the Serial1 on pins 19 (RX) and 18 (TX). For some other models see https://www.arduino.cc/en/Reference/Serial.

Please, let me know if this resolve your problem.

**Comment by jr-k** [comment]
I'm using Arduino Uno. My NFC reader is that one : http://www.banggood.com/fr/RC522-Chip-IC-Card-Induction-Module-RFID-Reader-p-81067.html

**Comment by MagoKimbra** [comment]
I could not get it to go, because it is set to work in i2c.
I have solved with this http://www.elechouse.com/elechouse/index.php?main_page=product_info&cPath=90_93&products_id=2258&zenid=lemnu9n477rsbvu951fqrmnf25

https://www.youtube.com/watch?v=-PMeuHj56-E

**Comment by yoonzy** [comment]
please help me zoider.. i has try simple example in the "examples" folder.. and i'm using MFRC522 with   ArduinoMega2560 I set the MISO(TX) pin to pins 19 (RX) and SDA(RX) pin to pins 18 (TX).. is this the correct configuration? I haven't got any response when opening the serial monitor..

**Comment by ricardojlrufino** [comment]
this works with this module ?
http://www.seekic.com/uploadfile/member_product/5199/20138242131767.jpg

**Comment by zodier** [comment]
Hi @yoonzy. Your configuration seems correct. But remember to change "&Serial" on the example code to "&Serial1" if you are using pins 19 (RX) and 18 (TX) with Arduino Mega.

Hi @ricardojlrufino. My module is not exactly that one ([is this](http://www.elechouse.com/elechouse/images/product/13.56MHZ_RFID_Module/5.jpg)), but you can try it and tell me if it works. :)

**Comment by ricardojlrufino** [comment]
I think that my module is for SPI only

**Comment by nguyentuancdt1** [comment]
Hi #zodier, i'm using mega2560 and rfid_module_v4 elechouse, i don't know the way to write a byte to card, i can write by SPI comm, can you teach me the way to write by uart, thank !

rfid module http://www.elechouse.com/elechouse/index.php?main_page=product_info&products_id=2156

**Comment by Himura2la** [comment]
I'm trying to understand how should I talk to this chip via UART and one evening was not enough for this. 
I've found a question. Where does the 0xAB came from? As I see in the datasheet, to read a register you should send zero, then 8 bit is data then one. So one the package is 10bit. And your code looks weird, I can't understand what do you send and why should it work. Also I don't see the serial number request, why do you expect the chip will send it on init?

**Comment by nguyentuancdt1** [comment]
Hi zodier !

 

I’m sorry because now I reply, I just don’t understand the way to send & receive module MFRC522 by UART, also if by SPI I can write the number to card very easy

Or read number of card again, but I think speed R/W SPI slower UART, can you teach me the way write/read data to card, example

 

Write number 1 to card, after that read card.

 

Now I have project make auto car, Car will full other car to transfer material. The auto car will read card to know position such as warehouse or station.

 

When I using SPI, sometime auto car pass card, so I think using uart will be better

 

Help me pls !

**Comment by Himura2la** [comment]
Use my repo, bro. I did it, my RC522 is working via UART. I found that the usual code that everybody use is quite weak, but as I googled, there's no alternatives((

**Comment by nguyentuancdt1** [comment]
Oh yes, I already see your code, now I have a mega2560 board, I’m using TX1/RX1(pin18,18) connect to MRFC522, TX0/RX0 to show serial but I don’t know change in your code to reconfig, I think have to change in  <https://github.com/zodier/MFRC522-UART-Arduino/blob/master/MFRC522.cpp> MFRC522.cpp and  <https://github.com/zodier/MFRC522-UART-Arduino/blob/master/MFRC522.h> MFRC522.h also Conector.begin(&Serial); allright ?

 

If I want to write number 1 into card, is it “Connector.write(1)”  ???

 

If I want to read and show the number of card on Serial0, is it “Serial.println(Connecto.read)”  ???

 

 

Can you teach me the way to change, thank bro !

**Comment by Himura2la** [comment]
Use the official repository https://github.com/miguelbalboa/rfid, this code is much more powerful.
First thing you have to do is rewrite the register-reading and -writing functions. 
https://github.com/miguelbalboa/rfid/blob/master/MFRC522.cpp#L46
https://github.com/miguelbalboa/rfid/blob/master/MFRC522.cpp#L61
https://github.com/miguelbalboa/rfid/blob/master/MFRC522.cpp#L79
https://github.com/miguelbalboa/rfid/blob/master/MFRC522.cpp#L95
Refer to my python code.

And watch all pin-dependant statements, you should correct them also.
like here https://github.com/miguelbalboa/rfid/blob/master/MFRC522.cpp#L199

And don't forger to switch your board to UART mode. http://igor-kochet.livejournal.com/267758.html

**Comment by nguyentuancdt1** [comment]
Exactly, I already use https://github.com/miguelbalboa/rfid this library to SPI comm, read and write data card normal, now I want to use UART comm but can not :(. I use MRFID-RC522 in elechouse it’s not RFID-RC522, it don’t have switch SPI-UART. Haizzz !!!!

 

Help me bro ^^

**Comment by Himura2la** [comment]
There's no 'switch'. you will need to cut the pull-up track to pin 32 as shown here http://igor-kochet.livejournal.com/267758.html
Also, I have to notice that the task you are going to achieve is quite complicated. You will need to read the datasheet, understand the existing code (and how it is connected to the datasheet) then write your own code accordng to the datasheet. Of corse now you also have my working code in another language, but it simplifies your task just a little bit. In case something should not work as expected, there will be noone to help except for the datasheet.
You can go on searching for a ready solution in case it's too hard for you.
Or you can drop your arduino and start using the real-world hardware and software. In this case you will have my ready code.

**Comment by nguyentuancdt1** [comment]
Thank  <https://github.com/Himura2la> Himura2la !

 

Does your code have connect between adruino uno and mrfid-rc522 ?

 

I will review again !

**Comment by Himura2la** [comment]
I don't use arduino. My code connects rc522 with a PC or anything that could execute Python. For instance, Raspberry Pi or ESP8266

**Comment by nguyentuancdt1** [comment]
oh oh, so i don't understand because your code differance adruino, sorry now i'm newbie hihi

**Comment by Himura2la** [comment]
That's why I warned you this is not an plug-and-flash task and may require time and knowledge...

**Comment by nguyentuancdt1** [comment]
ok thank you !
i will trying

**Comment by cyberjupiter** [comment]
@Himura2la can you write us a simple read and write for Arduino using UART? I would be glad even if it is not a complete code.

**Comment by Himura2la** [comment]
Sorry, I don't have time to translate my code into the language I hate ( 
Python is extremely readable, I hope you won't have problems in understanding my code

**Comment by cyberjupiter** [comment]
Okay sorry but maybe I was too demanding. I have never worked with Python but I will give it a shot. Would you mind if I contact you from time to time for a specific question regard with the code?

**Comment by Himura2la** [comment]
Sure. You definitely have to try Python, it's just perfect.

**Comment by Prasset** [comment]
i am so confused about  DTRQ and MX pins in UART communication. I cant found it in RC522 module.

**Comment by Himura2la** [comment]
They are not required, in fact I never ever used them and even don't know
their purpose.
UART is RX, TX and common GND
